src/blog/wasm-tail-call.md

```diff
@@ -0,0 +1,165 @@
+
+## Tail Calls in V8
+
+As we saw earlier, it is not the engine’s responsibility to detect calls in tail position. This should be done upstream by the toolchain. So the only thing left to do for TurboFan (V8's optimizing compiler) is to emit an appropriate sequence of instructions based on the call kind and the target function signature.  For our fibonacci example from earlier, the stack would look like this:
+
+![Simple tail call in TurboFan](/_img/tail-call.png)
+
+On the left we are inside `fib_rec` (green), called by `fib` (blue) and about to recursively tail-call `fib_rec`. First we unwind the current frame by resetting the frame and stack pointer. The frame pointer just restores its previous value by reading it from the “Caller FP” slot. The stack pointer moves to the top of the parent frame, plus enough space for any potential stack parameters and stack return values for the callee (0 in this case, everything is passed by registers). Parameters are moved into their expected registers according to fib_rec’s linkage (not shown in the diagram). And finally we start running `fib_rec`, which will start by creating a new frame.
+
+`fib_rec` will unwind and rewind itself like this until `n == 0`, at which point it will return `a` by register to `fib`.
+
+This is a simple case where all parameters and return values fit into registers, and the callee has the same signature as the caller. In the general case, we might need to do complex stack manipulations:
+
+- Read outgoing parameters from the old frame
+- Move parameters into the new frame
+- Adjust the frame size by moving the return address up or down, depending on the number of stack parameters in the callee
+
+All these reads and writes can conflict with each other, because we are reusing the same stack space. This is a crucial difference with a non-tail call, which would simply push all the stack parameters and the return address on top of the stack.
+
+![Complex tail call in TurboFan](/_img/tail-call-complex.png)
+
+TurboFan handles these stack and register manipulations with the “gap resolver”, a component which takes a list of moves that should semantically be executed in parallel, and generates the appropriate sequence of moves to resolve potential interferences between the move’s sources and destinations. If the conflicts are acyclic, this is just a matter of reordering the moves such that all sources are read before they are overwritten. For cyclic conflicts (e.g. if we swap two stack parameters), this can involve moving one of the sources to a temporary register or a temporary stack slot to break the cycle.
+
+Tail calls are also supported in Liftoff, our baseline compiler. However they are not optimized in this tier: Liftoff will push the parameters, return address and frame pointer to complete the frame as if this was a regular call, and then shift everything downwards to discard the caller frame:
```

Comment:
Should we mention that we *have to* implement tail calls properly in Liftoff, to avoid running out of stack space?

Reply:
Done. There is also the issue with tail-calls inside a try block, which wouldn't work as expected if we did regular calls in Liftoff, but I'm not sure we want to be so technical in this article...

Reply:
I don't think we need to be that specific. But saying that return calls must not be implemented via regular calls is an important point.